In [1]:
!pip install tensorflow keras numpy matplotlib pandas sklearn Bio

     |████████████████████████████████| 2.3MB 7.7MB/s 


### **Exploring Datasets with pandas**
pandas is an essential data analysis toolkit for Python. From their website:

pandas is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, real world data analysis in Python.

The course heavily relies on pandas for data wrangling, analysis, and visualization. We encourage you to spend some time and familizare yourself with the pandas API Reference: http://pandas.pydata.org/pandas-docs/stable/api.html.


**Datasets:**


1.   Immigration to Canada from 1980 to 2013 - [International migration flows to and from selected countries - The 2015 revision](https://www.un.org/en/development/desa/population/migration/data/empirical2/migrationflows.asp) from United Nation's website. The dataset contains annual data on the flows of international migrants as recorded by the countries of destination. The data presents both inflows and outflows according to the place of birth, citizenship or place of previous / next residence both for foreigners and nationals. For this lesson, we will focus on the Canadian Immigration data


### **Downloading and Prepping Data**
pandas Basics
The first thing we'll do is import two key data analysis modules: pandas and Numpy.

In [2]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library
from PIL import Image # converting images into arrays
%matplotlib inline

import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches # needed for waffle Charts

mpl.style.use('ggplot') # optional: for ggplot-like style

# check for latest version of Matplotlib
print ('Matplotlib version: ', mpl.__version__) # >= 2.0.0

Matplotlib version:  3.2.2


Download the dataset and read it into a pandas dataframe:

In [3]:
df_can = pd.read_excel('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DV0101EN/labs/Data_Files/Canada.xlsx',
                       sheet_name='Canada by Citizenship',
                       skiprows=range(20),
                       skipfooter=2)

print('Data downloaded and read into a dataframe!')

Data downloaded and read into a dataframe!


Let's take a look at the first five items in our dataset

In [4]:
df_can.head()
# tip: You can specify the number of rows you'd like to see as follows: df_can.head(10) 

Type    Coverage          OdName  AREA  ...  2010  2011  2012  2013
0  Immigrants  Foreigners     Afghanistan   935  ...  1758  2203  2635  2004
1  Immigrants  Foreigners         Albania   908  ...   561   539   620   603
2  Immigrants  Foreigners         Algeria   903  ...  4752  4325  3774  4331
3  Immigrants  Foreigners  American Samoa   909  ...     0     0     0     0
4  Immigrants  Foreigners         Andorra   908  ...     0     0     1     1

[5 rows x 43 columns]

When analyzing a dataset, it's always a good idea to start by getting basic information about your dataframe. We can do this by using the info() method.

In [5]:
df_can.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 43 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Type      195 non-null    object
 1   Coverage  195 non-null    object
 2   OdName    195 non-null    object
 3   AREA      195 non-null    int64 
 4   AreaName  195 non-null    object
 5   REG       195 non-null    int64 
 6   RegName   195 non-null    object
 7   DEV       195 non-null    int64 
 8   DevName   195 non-null    object
 9   1980      195 non-null    int64 
 10  1981      195 non-null    int64 
 11  1982      195 non-null    int64 
 12  1983      195 non-null    int64 
 13  1984      195 non-null    int64 
 14  1985      195 non-null    int64 
 15  1986      195 non-null    int64 
 16  1987      195 non-null    int64 
 17  1988      195 non-null    int64 
 18  1989      195 non-null    int64 
 19  1990      195 non-null    int64 
 20  1991      195 non-null    int64 
 21  1992      195 no

To get the list of column headers we can call upon the dataframe's .columns parameter.

In [6]:
df_can.columns.values 

array(['Type', 'Coverage', 'OdName', 'AREA', 'AreaName', 'REG', 'RegName',
       'DEV', 'DevName', 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987,
       1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998,
       1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009,
       2010, 2011, 2012, 2013], dtype=object)

Similarly, to get the list of indicies we use the .index parameter.

In [7]:
df_can.index.values

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

Note: The default type of index and columns is NOT list.

In [8]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


To get the index and columns as lists, we can use the tolist() method.

In [9]:
df_can.columns.tolist()
df_can.index.tolist()

print (type(df_can.columns.tolist()))
print (type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


Let's find out how many entries there are in our dataset

In [10]:
# print the dimensions of the dataframe
print(df_can.shape)

(195, 43)


**Cleaning up data:**

Let's clean the data set to remove a few unnecessary columns. We can use pandas drop() method as follows:

In [11]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

OdName AreaName          RegName  ...  2011  2012  2013
0  Afghanistan     Asia    Southern Asia  ...  2203  2635  2004
1      Albania   Europe  Southern Europe  ...   539   620   603

[2 rows x 38 columns]

Let's rename the columns so that they make sense. We can use rename() method by passing in a dictionary of old and new names as follows:

In [12]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

Index([  'Country', 'Continent',    'Region',   'DevName',        1980,
              1981,        1982,        1983,        1984,        1985,
              1986,        1987,        1988,        1989,        1990,
              1991,        1992,        1993,        1994,        1995,
              1996,        1997,        1998,        1999,        2000,
              2001,        2002,        2003,        2004,        2005,
              2006,        2007,        2008,        2009,        2010,
              2011,        2012,        2013],
      dtype='object')

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:

In [13]:
df_can['Total'] = df_can.sum(axis=1)

We can check to see how many null objects we have in the dataset as follows:

In [14]:
df_can.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Finally, let's view a quick summary of each column in our dataframe using the describe() method.

In [15]:
df_can.describe()

1980          1981  ...          2013          Total
count    195.000000    195.000000  ...    195.000000     195.000000
mean     508.394872    566.989744  ...   1320.702564   32867.451282
std     1949.588546   2152.643752  ...   4237.951988   91785.498686
min        0.000000      0.000000  ...      0.000000       1.000000
25%        0.000000      0.000000  ...     45.000000     952.000000
50%       13.000000     10.000000  ...    213.000000    5018.000000
75%      251.500000    295.500000  ...    796.000000   22239.500000
max    22045.000000  24796.000000  ...  34129.000000  691904.000000

[8 rows x 35 columns]

### **pandas Intermediate: Indexing and Selection (slicing)**
**Select Column**

There are two ways to filter on a column name:

**Method 1:** Quick and easy, but only works if the column name does NOT have spaces or special characters.

    df.column_name 
        (returns series)
**Method 2:** More robust, and can filter on multiple columns.

    df['column']  
        (returns series)
    df[['column 1', 'column 2']] 
        (returns dataframe)

Example: Let's try filtering on the list of countries ('Country').

In [16]:
df_can.Country  # returns a series

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Let's try filtering on the list of countries ('OdName') and the data for years: 1980 - 1985.

In [17]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe
# notice that 'Country' is string, and the years are integers. 
# for the sake of consistency, we will convert all column names to string later on.

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### **Select Row**
There are main 3 ways to select rows:

    df.loc[label]        
        #filters by the labels of the index/column
    df.iloc[index]       
        #filters by the positions of the index/column
Before we proceed, notice that the defaul index of the dataset is a numeric range from 0 to 194. This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corressponding index value.

This can be fixed very easily by setting the 'Country' column as the index using set_index() method.

In [18]:
df_can.set_index('Country', inplace=True)
# tip: The opposite of set is reset. So to reset the index, we can use df_can.reset_index()

In [19]:
df_can.head(3)

Continent           Region             DevName  ...  2012  2013  Total
Country                                                     ...                   
Afghanistan      Asia    Southern Asia  Developing regions  ...  2635  2004  58639
Albania        Europe  Southern Europe   Developed regions  ...   620   603  15699
Algeria        Africa  Northern Africa  Developing regions  ...  3774  4331  69439

[3 rows x 38 columns]

In [20]:
#  to remove the name of the index
df_can.index.name = None

Example: Let's view the number of immigrants from Japan (row 87) for the following scenarios: 1. The full row data (all columns) 2. For year 2013 3. For years 1980 to 1985

In [21]:
# 1. the full row data (all columns)
print(df_can.loc['Japan'])

# alternate methods
print(df_can.iloc[87])
print(df_can[df_can.index == 'Japan'].T.squeeze())

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [22]:
# 2. for year 2013
print(df_can.loc['Japan', 2013])

# alternate method
print(df_can.iloc[87, 36]) # year 2013 is the last column, with a positional index of 36

982
982


In [23]:
# 3. for years 1980 to 1985
print(df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1985]])
print(df_can.iloc[87, [3, 4, 5, 6, 7, 8]])

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object
1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object


Column names that are integers (such as the years) might introduce some confusion. For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index.

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.

In [24]:
df_can.columns = list(map(str, df_can.columns))
# [print (type(x)) for x in df_can.columns.values] #<-- uncomment to check type of column headers

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:


In [25]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

### **Filtering based on a criteria**

To filter the dataframe based on a condition, we simply pass the condition as a boolean vector.

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).

In [26]:
# 1. create the condition boolean series
condition = df_can['Continent'] == 'Asia'
print(condition)

Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [27]:
# 2. pass this condition into the dataFrame
df_can[condition]

Continent  ...   Total
Afghanistan                                         Asia  ...   58639
Armenia                                             Asia  ...    3310
Azerbaijan                                          Asia  ...    2649
Bahrain                                             Asia  ...     475
Bangladesh                                          Asia  ...   65568
Bhutan                                              Asia  ...    5876
Brunei Darussalam                                   Asia  ...     600
Cambodia                                            Asia  ...    6538
China                                               Asia  ...  659962
China, Hong Kong Special Administrative Region      Asia  ...    9327
China, Macao Special Administrative Region          Asia  ...     284
Cyprus                                              Asia  ...    1126
Democratic People's Republic of Korea               Asia  ...     388
Georgia                                             Asia  ...    2068
India                                               Asia  ...  691904
Indonesia                                           Asia  ...   13150
Iran (Islamic Republic of)                          Asia  ...  175923
Iraq                                                Asia  ...   69789
Israel                                              Asia  ...   66508
Japan                                               Asia  ...   27707
Jordan                                              Asia  ...   35406
Kazakhstan                                          Asia  ...    8490
Kuwait                                              Asia  ...    2025
Kyrgyzstan                                          Asia  ...    2353
Lao People's Democratic Republic                    Asia  ...    1089
Lebanon                                             Asia  ...  115359
Malaysia                                            Asia  ...   24417
Maldives                                            Asia  ...      30
Mongolia                                            Asia  ...     952
Myanmar                                             Asia  ...    9245
Nepal                                               Asia  ...   10222
Oman                                                Asia  ...     224
Pakistan                                            Asia  ...  241600
Philippines                                         Asia  ...  511391
Qatar                                               Asia  ...     157
Republic of Korea                                   Asia  ...  142581
Saudi Arabia                                        Asia  ...    3425
Singapore                                           Asia  ...   14579
Sri Lanka                                           Asia  ...  148358
State of Palestine                                  Asia  ...    6512
Syrian Arab Republic                                Asia  ...   31485
Tajikistan                                          Asia  ...     503
Thailand                                            Asia  ...    9174
Turkey                                              Asia  ...   31781
Turkmenistan                                        Asia  ...     310
United Arab Emirates                                Asia  ...     836
Uzbekistan                                          Asia  ...    3368
Viet Nam                                            Asia  ...   97146
Yemen                                               Asia  ...    2985

[49 rows x 38 columns]

In [28]:
# we can pass mutliple criteria in the same line. 
# let's filter for AreaNAme = Asia and RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

Continent         Region  ...   2013   Total
Afghanistan                     Asia  Southern Asia  ...   2004   58639
Bangladesh                      Asia  Southern Asia  ...   3789   65568
Bhutan                          Asia  Southern Asia  ...    487    5876
India                           Asia  Southern Asia  ...  33087  691904
Iran (Islamic Republic of)      Asia  Southern Asia  ...  11291  175923
Maldives                        Asia  Southern Asia  ...      1      30
Nepal                           Asia  Southern Asia  ...   1308   10222
Pakistan                        Asia  Southern Asia  ...  12603  241600
Sri Lanka                       Asia  Southern Asia  ...   2394  148358

[9 rows x 38 columns]

Before we proceed: let's review the changes we have made to our dataframe.

In [29]:
print('data dimensions:', df_can.shape)
print(df_can.columns)
df_can.head(2)

data dimensions: (195, 38)
Index(['Continent', 'Region', 'DevName', '1980', '1981', '1982', '1983',
       '1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991', '1992',
       '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001',
       '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009', '2010',
       '2011', '2012', '2013', 'Total'],
      dtype='object')


Continent           Region             DevName  ...  2012  2013  Total
Afghanistan      Asia    Southern Asia  Developing regions  ...  2635  2004  58639
Albania        Europe  Southern Europe   Developed regions  ...   620   603  15699

[2 rows x 38 columns]

Thank you ...! This notebook is part of a course on Coursera called Data Visualization with Python. If you accessed this notebook outside the course, you can take this course online by clicking [here](https://www.coursera.org/learn/python-for-data-visualization).